--- linux-6.1.0/arch/powerpc/platforms/83xx/mpc832x_rdb.c	2023-07-05 08:37:14.869303500 +0900
+++ linux-xlnx-2023.1/arch/powerpc/platforms/83xx/mpc832x_rdb.c	2023-07-05 08:33:12.656949200 +0900
@@ -144,7 +144,7 @@
 
 static void mpc83xx_spi_cs_control(struct spi_device *spi, bool on)
 {
-	pr_debug("%s %d %d\n", __func__, spi->chip_select, on);
+	pr_debug("%s %d %d\n", __func__, spi_get_chipselect(spi, 0), on);
 	par_io_data_set(3, 13, on);
 }
 
