--- linux-6.1.0/drivers/bluetooth/hci_ll.c	2023-07-05 08:37:18.723879800 +0900
+++ linux-xlnx-2023.1/drivers/bluetooth/hci_ll.c	2023-07-05 08:33:15.512022600 +0900
@@ -626,7 +626,7 @@
 		gpiod_set_value_cansleep(lldev->enable_gpio, 0);
 		msleep(5);
 		gpiod_set_value_cansleep(lldev->enable_gpio, 1);
-		mdelay(100);
+		msleep(500);
 		err = serdev_device_wait_for_cts(serdev, true, 200);
 		if (err) {
 			bt_dev_err(hu->hdev, "Failed to get CTS");
