--- linux-6.1.0/drivers/iommu/iommu.c	2023-07-05 08:37:26.065718000 +0900
+++ linux-xlnx-2023.1/drivers/iommu/iommu.c	2023-07-05 08:33:22.928287900 +0900
@@ -27,6 +27,7 @@
 #include <linux/fsl/mc.h>
 #include <linux/module.h>
 #include <linux/cc_platform.h>
+#include <linux/cdx/cdx_bus.h>
 #include <trace/events/iommu.h>
 
 #include "dma-iommu.h"
@@ -123,6 +124,9 @@
 #ifdef CONFIG_TEGRA_HOST1X_CONTEXT_BUS
 	&host1x_context_device_bus_type,
 #endif
+#ifdef CONFIG_CDX_BUS
+	&cdx_bus_type,
+#endif
 };
 
 /*
