--- linux-6.1.0/drivers/media/test-drivers/vivid/vivid-vid-cap.c	2023-07-05 08:37:27.638320200 +0900
+++ linux-xlnx-2023.1/drivers/media/test-drivers/vivid/vivid-vid-cap.c	2023-07-05 08:33:24.483056400 +0900
@@ -1907,7 +1907,7 @@
 			break;
 	if (i == ARRAY_SIZE(webcam_sizes))
 		return -EINVAL;
-	if (fival->index >= 2 * (VIVID_WEBCAM_SIZES - i))
+	if (fival->index >= 2 * VIVID_WEBCAM_SIZES)
 		return -EINVAL;
 	fival->type = V4L2_FRMIVAL_TYPE_DISCRETE;
 	fival->discrete = webcam_intervals[fival->index];
@@ -1934,7 +1934,7 @@
 			  struct v4l2_streamparm *parm)
 {
 	struct vivid_dev *dev = video_drvdata(file);
-	unsigned ival_sz = 2 * (VIVID_WEBCAM_SIZES - dev->webcam_size_idx);
+	unsigned ival_sz = 2 * VIVID_WEBCAM_SIZES;
 	struct v4l2_fract tpf;
 	unsigned i;
 
--- linux-6.1.0/drivers/media/test-drivers/vivid/vivid-vid-common.c	2023-07-05 08:37:27.638320200 +0900
+++ linux-xlnx-2023.1/drivers/media/test-drivers/vivid/vivid-vid-common.c	2023-07-05 08:33:24.483056400 +0900
@@ -74,6 +74,14 @@
 		.buffers = 1,
 	},
 	{
+		.fourcc   = V4L2_PIX_FMT_YUV444P,
+		.vdownsampling = { 1, 1, 1 },
+		.bit_depth = { 8, 8, 8 },
+		.color_enc = TGP_COLOR_ENC_YCBCR,
+		.planes   = 3,
+		.buffers = 1,
+	},
+	{
 		.fourcc   = V4L2_PIX_FMT_YUV420,
 		.vdownsampling = { 1, 2, 2 },
 		.bit_depth = { 8, 4, 4 },
