--- linux-6.1.0/Documentation/devicetree/bindings/mfd/syscon.yaml	2023-07-05 08:37:05.879666500 +0900
+++ linux-xlnx-2023.1/Documentation/devicetree/bindings/mfd/syscon.yaml	2023-07-05 08:33:03.934341500 +0900
@@ -69,6 +69,7 @@
               - samsung,exynos5433-sysreg
               - samsung,exynos850-sysreg
               - samsung,exynosautov9-sysreg
+              - xlnx,zynqmp-dpaud-setting
 
           - const: syscon
 
--- linux-6.1.0/include/linux/mfd/syscon/xlnx-vcu.h	2023-07-05 08:37:38.979156300 +0900
+++ linux-xlnx-2023.1/include/linux/mfd/syscon/xlnx-vcu.h	2023-07-05 08:33:36.075346600 +0900
@@ -36,4 +36,32 @@
 #define VCU_GASKET_INIT			0x74
 #define VCU_GASKET_VALUE		0x03
 
+/**
+ * struct xvcu_device - Xilinx VCU init device structure
+ * @dev: Platform device
+ * @pll_ref: pll ref clock source
+ * @reset_gpio: vcu reset gpio
+ * @aclk: axi clock source
+ * @logicore_reg_ba: logicore reg base address
+ * @vcu_slcr_ba: vcu_slcr Register base address
+ * @pll: handle for the VCU PLL
+ * @pll_post: handle for the VCU PLL post divider
+ * @clk_data: clocks provided by the vcu clock provider
+ */
+struct xvcu_device {
+	struct device *dev;
+	struct clk *pll_ref;
+	struct clk *aclk;
+	struct gpio_desc *reset_gpio;
+	struct regmap *logicore_reg_ba;
+	void __iomem *vcu_slcr_ba;
+	struct clk_hw *pll;
+	struct clk_hw *pll_post;
+	struct clk_hw_onecell_data *clk_data;
+};
+
+u32 xvcu_get_color_depth(struct xvcu_device *xvcu);
+u32 xvcu_get_memory_depth(struct xvcu_device *xvcu);
+u32 xvcu_get_clock_frequency(struct xvcu_device *xvcu);
+u32 xvcu_get_num_cores(struct xvcu_device *xvcu);
 #endif /* __XLNX_VCU_H */
