--- linux-6.1.0/drivers/usb/storage/unusual_uas.h	2023-07-05 08:37:35.949089100 +0900
+++ linux-xlnx-2023.1/drivers/usb/storage/unusual_uas.h	2023-07-05 08:33:33.048028400 +0900
@@ -76,6 +76,12 @@
 		USB_SC_DEVICE, USB_PR_DEVICE, NULL,
 		US_FL_IGNORE_UAS),
 
+UNUSUAL_DEV(0x0525, 0xa4a5, 0x0000, 0x9999,
+	    "Netchip",
+		"Target Product",
+		USB_SC_DEVICE, USB_PR_DEVICE, NULL,
+		US_FL_NO_ATA_1X),
+
 /* Reported-by: David Webb <djw@noc.ac.uk> */
 UNUSUAL_DEV(0x0bc2, 0x331a, 0x0000, 0x9999,
 		"Seagate",
